### PR TITLE
CMake: fixed detection and use of iso9660++ library

### PR DESCRIPTION
**Pull request by Karlson2k**
This is a standalone fix-only subset of #19197.

It doesn't change anything except correct detection of `iso9660++` and correct library order for `iso9660++`.

As it's minimal, I believe it's fine for v19.

### CODE REVIEW

File: cmake/modules/FindIso9660pp.cmake [hunk 1]
```diff
@@ -15,14 +15,16 @@ if(PKG_CONFIG_FOUND)
   pkg_check_modules(PC_ISO9660 libiso9660>=2.1.0 QUIET)
 endif()
 
+find_package(Cdio)
```
Comment (wsnipex) [inline cmake/modules/FindIso9660pp.cmake:18]:
```suggestion
find_package(Cdio REQUIRED)
```

File: cmake/modules/FindIso9660pp.cmake [hunk 2]
```diff
@@ -15,14 +15,16 @@ if(PKG_CONFIG_FOUND)
...
+
 find_path(ISO9660PP_INCLUDE_DIR NAMES cdio++/iso9660.hpp
-                                PATHS ${PC_ISO9660PP_INCLUDEDIR})
+                                PATHS ${PC_ISO9660PP_INCLUDEDIR} ${CDIO_INCLUDE_DIR})
```
Comment (wsnipex) [inline cmake/modules/FindIso9660pp.cmake:21]: not needed

File: cmake/modules/FindIso9660pp.cmake [hunk 3]
```diff
@@ -15,14 +15,16 @@ if(PKG_CONFIG_FOUND)
...
 
 find_library(ISO9660PP_LIBRARY NAMES libiso9660++ iso9660++
                                PATHS ${PC_ISO9660PP_LIBDIR})
 
 find_path(ISO9660_INCLUDE_DIR NAMES cdio/iso9660.h
-                              PATHS ${PC_ISO9660_INCLUDEDIR})
+                              PATHS ${PC_ISO9660_INCLUDEDIR} ${CDIO_INCLUDE_DIR})
```
Comment (wsnipex) [inline cmake/modules/FindIso9660pp.cmake:27]: not needed

File: cmake/modules/FindIso9660pp.cmake [hunk 4]
```diff
@@ -31,12 +33,12 @@ set(ISO9660PP_VERSION ${PC_ISO9660PP_VERSION})
 
 include(FindPackageHandleStandardArgs)
 find_package_handle_standard_args(Iso9660pp
-                                  REQUIRED_VARS ISO9660PP_LIBRARY ISO9660PP_INCLUDE_DIR ISO9660_LIBRARY ISO9660_INCLUDE_DIR
+                                  REQUIRED_VARS ISO9660PP_LIBRARY ISO9660PP_INCLUDE_DIR CDIOPP_INCLUDE_DIR ISO9660_LIBRARY ISO9660_INCLUDE_DIR CDIO_LIBRARY CDIO_INCLUDE_DIR
```
Comment (wsnipex) [inline cmake/modules/FindIso9660pp.cmake:36]: not needed

**Review by wsnipex — CHANGES_REQUESTED** [review]
libiso9660 must be a required dep if ENABLE_OPTICAL is true

**Comment by Karlson2k** [comment]
> libiso9660 must be a required dep if ENABLE_OPTICAL is true

@wsnipex, actually it is a separate feature.
`libcdio` is used to control optical drives (load/eject tray, read audio CD etc.), while `iso9660++` is used to read iso images (from files or from devices), like `libudfread` is used to read udf images.

As an example, Kodi is currently compiled for Windows and Ubuntu without `iso9660`, but with `libcdio`.

It is separated in the code.

File: cmake/modules/FindCdio.cmake [hunk 5]
```diff
@@ -19,6 +20,14 @@ find_path(CDIO_INCLUDE_DIR NAMES cdio/cdio.h
...
                           PATHS ${PC_CDIO_LIBDIR})
 
+# cdio++ header is not used directly by Kodi, but it's required for iso9660++ lib
+find_path(CDIOPP_INCLUDE_DIR NAMES cdio++/cdio.hpp
+                             PATHS ${PC_CDIOPP_INCLUDEDIR} ${CDIO_INCLUDE_DIR})
+
+# Currently nothing requires libcdio++ binary, but situation can change with future version of iso9660++
+#find_library(CDIOPP_LIBRARY NAMES libcdio++ cdio++
+#                            PATHS ${PC_CDIOPP_LIBDIR})
+
```
[the remark this answers is not in the document]
Comment (Karlson2k) [reply cmake/modules/FindCdio.cmake:30]: As it is a separate feature, it must be included in `iso9660++`, but not in `libcdio`.

Maybe I should move it back (to `FindIso9660`), like I implemented it originally.

**Comment by wsnipex** [comment]
it's not really seperate: iso9660++ requires cdio, so if we want ISO playback, we need iso9660++ AND libcdio.

File: cmake/modules/FindIso9660pp.cmake (hunk 2 above)
Comment (Karlson2k) [reply cmake/modules/FindIso9660pp.cmake:21]: If `libcdio` where found, it makes sense to add know path to the list, as `PATHS` are just a hints for CMake search.

File: cmake/modules/FindIso9660pp.cmake (hunk 3 above)
Comment (Karlson2k) [reply cmake/modules/FindIso9660pp.cmake:27]: The same. May help to find the lib.

File: cmake/modules/FindIso9660pp.cmake (hunk 4 above)
Comment (Karlson2k) [reply cmake/modules/FindIso9660pp.cmake:36]: Required to be here.

`cdio++` is needed if and only if `iso9600++` is used.

**Comment by Karlson2k** [comment]
While `iso9660++` require `libcdio`, but `libcdio` can be used without `iso9660++`.
That's why I think that my original implementation is better: `libcdio` is detected without `libcdio++` as `libcdio++` is not needed for `libcdio`.
Detection module for `iso9660++` should detect `cdio++`.

**Comment by Karlson2k** [comment]
> sorry, that should probably be: cdio must be a required dep if ENABLE_OPTICAL is true

@wsnipex Absolutely. But currently `libcdio` is hard dependency of Kodi. See #19197.

**Comment by wsnipex** [comment]
don't make it too complicated. cdio provides cdio++ (if high enough version, which we require), so we can simply fail if cdio is missing the cpp headers.

File: cmake/modules/FindIso9660pp.cmake (hunk 1 above)
Comment (Karlson2k) [reply cmake/modules/FindIso9660pp.cmake:18]: As `iso9660` is a separate optional feature and not required by many Kodi installations, it break ability to build Kodi without `iso9660` lib.

File: cmake/modules/FindCdio.cmake [hunk 6]
```diff
@@ -11,6 +11,7 @@
 
 if(PKG_CONFIG_FOUND)
   pkg_check_modules(PC_CDIO libcdio>=0.80 QUIET)
```
Comment (wsnipex) [inline cmake/modules/FindCdio.cmake:13]:
```suggestion
  pkg_check_modules(PC_CDIO libcdio>=2.1.0 QUIET)
```

**Comment by Karlson2k** [comment]
> > While iso9660++ require libcdio, but libcdio can be used without iso9660++.
> > That's why I think that my original implementation is better: libcdio is detected without libcdio++ as libcdio++ is not needed for libcdio.
> > Detection module for iso9660++ should detect cdio++.
> 
> don't make it too complicated. cdio provides cdio++ (if high enough version, which we require), so we can simply fail if cdio is missing the cpp headers.

I believe it's not the best approach. Why to fail if missing headers that are not required.
Kodi is build on many platforms, not only Ubuntu.
On Gentoo, for example, libcdio can be built without c++ interface. On other platforms, libcdio can be present but without `iso9660++`.
It's much better to use what is available, then block feature.

**Comment by Karlson2k** [comment]
Everything is handled correctly in #19197.

**Comment by wsnipex** [comment]
no, it is not correctly handled.
- There is no check if cdio.hpp is actually available
- iso9660++ is required for ISO playback, therefore we also need cdio++ and cdio
- don't mix and match Find modules. Everything cdio related should be handled in FindCdio.cmake

File: cmake/modules/FindCdio.cmake (hunk 6 above)
Comment (Karlson2k) [reply cmake/modules/FindCdio.cmake:13]: Is `libcdio` below 2.1.0 really bad? 
Current and previous Debian, Fedora 32 and many other distros doesn't have 2.1.0

**Comment by Karlson2k** [comment]
It's good that we've started with concrete points.

> * There is no check if cdio.hpp is actually available

Is it here: https://github.com/xbmc/xbmc/pull/19230/files#diff-15e3ebf81ac428dd8ec52f4fe81cac4ff13736da4f7f07a8a8e19cb057d6b614R24
And here: https://github.com/xbmc/xbmc/pull/19230/files#diff-88004735dc842191b2b75b219193c7c288422f0cfa1d567ed0dc2519f3db7695R36

> * iso9660++ is required for ISO playback, therefore we also need cdio++ and cdio

`iso9660++` can be enabled only if all required dependencies are found: https://github.com/xbmc/xbmc/pull/19230/files#diff-88004735dc842191b2b75b219193c7c288422f0cfa1d567ed0dc2519f3db7695R36

> * don't mix and match Find modules. Everything cdio related should be handled in FindCdio.cmake

Should I move `cdio++` detection to the separate module `FindCdiopp.cmake`?

File: cmake/modules/FindCdio.cmake (hunk 6 above)
Comment (wsnipex) [reply cmake/modules/FindCdio.cmake:13]: I repeat: cdio++ is provided by cdio, so we must use the same version. If you manage to install 0.80 and 2.1.0 in parallel, you do something seriously wrong.

**Comment by wsnipex** [comment]
> > * There is no check if cdio.hpp is actually available
> 
> Is it here: https://github.com/xbmc/xbmc/pull/19230/files#diff-15e3ebf81ac428dd8ec52f4fe81cac4ff13736da4f7f07a8a8e19cb057d6b614R24
> And here: https://github.com/xbmc/xbmc/pull/19230/files#diff-88004735dc842191b2b75b219193c7c288422f0cfa1d567ed0dc2519f3db7695R36

that is the wrong place as already pointed out several times. This belongs in FindCdio.cmake

> > * don't mix and match Find modules. Everything cdio related should be handled in FindCdio.cmake
> 
> Should I move `cdio++` detection to the separate module `FindCdiopp.cmake`?
no, it is part of cdio

**Comment by Karlson2k** [comment]
> > > * There is no check if cdio.hpp is actually available
> > 
> > 
> > Is it here: https://github.com/xbmc/xbmc/pull/19230/files#diff-15e3ebf81ac428dd8ec52f4fe81cac4ff13736da4f7f07a8a8e19cb057d6b614R24
> > And here: https://github.com/xbmc/xbmc/pull/19230/files#diff-88004735dc842191b2b75b219193c7c288422f0cfa1d567ed0dc2519f3db7695R36
> 
> that is the wrong place as already pointed out several times. This belongs in FindCdio.cmake
> 
> > > * don't mix and match Find modules. Everything cdio related should be handled in FindCdio.cmake
> > 
> > 
> > Should I move `cdio++` detection to the separate module `FindCdiopp.cmake`?
> 
> no, it is part of cdio

@wsnipex, sorry. Didn't get you.

Two question:
1. `cdio++` detection is in `FindCdio.cmake`. It is OK for you?
2. `cdio++` is not required to use `cdio` and usually provided in a separate package. Why cmake should fail for `cdio` if `cdio++` is not present?

**Comment by wsnipex** [comment]
ad 1.) mostly, see comments above
ad 2.) because it makes debugging issue due to silently missing deps a nightmare, see e.g. https://github.com/xbmc/xbmc/issues/18896

IMHO, iso playback is important enough to either make it required OR mark it as optional _explicitly_ by giving it it's own ENABLE option

**Comment by Karlson2k** [comment]
@wsnipex It already has a separate option.
CMake option is called `ENABLE_ISO9660PP` is it is `AUTO` by default.
Result of `iso9660++` detection is reported explicitly at the end or CMake output.
Precompiler macro `HAS_ISO9660PP` is defined if `iso9660++` is detected.

**Comment by wsnipex** [comment]
yes, but everything is AUTO, that's not explicit. Do you think that average joe even looks at the cmake output, let alone know that 
"iso9660++ not found" results in ISO playback being broken?

**Comment by Karlson2k** [comment]
The idea of this PR is to correctly detect all `iso9660++` dependencies and include all required libs to the list of libraries.

Currently presence of `iso9660++` can be detected even if `libcdio++` headers are missing. CMake enables `iso9660++` support and fails on the build stage.
This PR fixes it.
Is addition, this PR adds **all** required `iso9660++` binary dependencies in correct order.
That's the idea.

**Comment by Karlson2k** [comment]
> yes, but everything is AUTO, that's not explicit. Do you think that average joe even looks at the cmake output, let alone know that
> "iso9660++ not found" results in ISO playback being broken?

I suggest to move it to another PR.
This PR is a small fix for correct detection if `iso9660` which is currently broken.

CMake reporting and making something more explicit should be handled separately. And most probably will not be merged into v19.

**Comment by wsnipex** [comment]
>CMake reporting and making something more explicit should be handled separately. And most probably will not be merged into v19.

agreed. But that doesn't make the PR - in it's current state - correct.
I expect:
- libcdio and libcdio++ to be of the same version
- FindCdio.cmake to fail if cdio is enabled, but the cpp headers are missing

Since you don't seem to agree, let's leave it to @DaveTBlake to decide for v19 inclusion

**Comment by Karlson2k** [comment]
@wsnipex Let's find a consensus. :)

> * libcdio and libcdio++ to be of the same version

Good point, to avoid run-time problems. Could be handled by comparing `ISO9660PP_VERSION` and `CDIO_VERSION` with warning on version mismatch. This will attract more attention and still will allow to use older version of `libcdio` without `iso9660++`

> * FindCdio.cmake to fail if cdio is enabled, but the cpp headers are missing

Failing on missing cpp headers for `libcdio` effectively make Kodi unusable without headers which **are not** required for `libcdio`.
`iso9660++` detection should fail without cpp headers, while `libcdio` should succeed, right?

**Comment by Karlson2k** [comment]
Added check for version match.

The only point left is `fail libcdio if libcdo++ headers are not present`.
Let's talk about it.

If we will require it than Kodi will have hard dependency on `libcdio++`, while still can be built without `iso9660++` which is optional dependency.
In other words, we will require library, which will not be used. I think it's not the best way.

Alternatively, we can set by default `ENABLE_ISO9660PP` to `ON`, so "average Joe" will notice that something important is missing. It will be more efficient that hard dependency on `libcdio++`.

@wsnipex, what do you think?

**Comment by Karlson2k** [comment]
To summarize. The discrepancy:
* @wsnipex thinks that `libcdio` detection must fail if `libcdio++` headers are missing so user will notice it.
* I think that `libcdio` should fail only if `libcdio` itself is missing. As `libcdio++` header is used only for `iso9660++`, only `iso9660++` detection should fail if `libcdio++` header is missing.

Is it correct, @wsnipex?

**Comment by yol** [comment]
Sorry forgot to push the comment button...

@wsnipex Usually I'd agree and just say we combine libcdio, libcdio++, and maybe even libiso9660++ under one module and be done with it. But I see one big problem with this approach. Ubuntu 18.04 (as an example) only has libcdio 1.0.0. If we cobble everything together and always require libcdio++ together with libcdio, we will lose optical drive support on Ubuntu 18.04 because dependencies cannot be satisfied, correct?

**Comment by Karlson2k** [comment]
> @wsnipex Usually I'd agree and just say we combine libcdio, libcdio++, and maybe even libiso9660++ under one module and be done with it. But I see one big problem with this approach. Ubuntu 18.04 (as an example) only has libcdio 1.0.0. If we cobble everything together and always require libcdio++ together with libcdio, we will lose optical drive support on Ubuntu 18.04 because dependencies cannot be satisfied, correct?

That's correct.
With hard dependency on unneeded headers we loose more that earn.

The better version of this PR is #19234.

**Comment by wsnipex** [comment]
> Sorry forgot to push the comment button...
> 
> @wsnipex Usually I'd agree and just say we combine libcdio, libcdio++, and maybe even libiso9660++ under one module and be done with it. But I see one big problem with this approach. Ubuntu 18.04 (as an example) only has libcdio 1.0.0. If we cobble everything together and always require libcdio++ together with libcdio, we will lose optical drive support on Ubuntu 18.04 because dependencies cannot be satisfied, correct?

no, I've backported libcdio 2.1.0 to ubuntu 18.04 and 20.04: https://launchpad.net/~team-xbmc/+archive/ubuntu/xbmc-nightly/+packages?field.name_filter=libcdio&field.status_filter=published&field.series_filter=

**Comment by yol** [comment]
Honest question, no cynicism intended: What about the distro official packages? Or do we usually not care about those?

**Comment by wsnipex** [comment]
official v19 packages with be fine, since all major distros already have a new enough version, at least in their newest series.
E.g. ubuntu >= 20.10, fedora 32, gentoo, arch all have cdio >= 2.0

**Comment by yol** [comment]
so basically you say that newest versions' distro packages should be fine while LTS users have to use PPA?

**Comment by wsnipex** [comment]
on ubuntu, everyone should use the PPA.
there is no v19 on official repos.

**Comment by yol** [comment]
well, in the end, guess you're going to have to run an up-to-date distro for an up-to-date media center. So I'd be fine to just require newer libcdio and consolidate the features to get out of this mess... does that make sense?

**Comment by wsnipex** [comment]
yup, that's what I've been about all along

**Comment by Karlson2k** [comment]
So we are talking about making everything mandatory, right?
Obviously, it's another discussion, not really related to correct detection.

Let's decide about detection at first.
What we have:
1. Serval libs produced from libcdio sources: `libcdio`, `libcdio++`, `libiso9660`, `libiso9660++`, and `libudf`
2. One Kodi feature depends on `libcdio` lib.
3. Anther Kodi feature depends on `libiso9660++` (directly) and (indirectly) on `libiso9660++`, `libcdio++, and `libcdio`

What's the point to intermix dependencies and require `libcdo++` for the first Kodi feature?
If the idea is to implicitly require `libiso9660++` for basic Kodi build, then it fails because user still can install `libcdio++` but not install `libiso9660++`.

Such suggestion will only increase the mess.

Let's open a separate discussion about making something else mandatory and fix here only right dependencies.
Keep in mind that there are many distributions, with various set of libraries in their packages. You will not backport everything everywhere.

`libcdio++` is not required to enable `libcdio` functionality in Kodi. That's it.

**Comment by Karlson2k** [comment]
Superseded by #19234

File: cmake/modules/FindCdio.cmake (hunk 5 above)
[the remark this answers is not in the document]
Comment (Karlson2k) [reply cmake/modules/FindCdio.cmake:30]: As pointed by @yol, `cdio++` must be handled by distribution package dependencies. We are not making similar checks for other libraries as checking all included headers is senseless and would overcomplicate detection.

File: cmake/modules/FindIso9660pp.cmake (hunk 2 above)
Comment (Karlson2k) [reply cmake/modules/FindIso9660pp.cmake:21]: Removed

File: cmake/modules/FindIso9660pp.cmake (hunk 4 above)
Comment (Karlson2k) [reply cmake/modules/FindIso9660pp.cmake]: Removed

File: cmake/modules/FindCdio.cmake (hunk 6 above)
Comment (Karlson2k) [reply cmake/modules/FindCdio.cmake:13]: Removed

**Pull request closed without merging**